### PR TITLE
[FLINK-1758][gelly] Neighborhood Methods Extensions

### PR DESCRIPTION
**Pull request by andralungu**

### CODE REVIEW

**Comment by vasia** [comment]
Thanks for the PR @andralungu!

The idea for this issue is to (a) use a combinable reduce for the simple (without vertex value) reduceOn* methods, because it will be more efficient and (b) offer groupReduceOn* versions with Collectors to allow returning more than one value per vertex.
I see you have implemented (b) here, but not (a).

To be more specific, the groupReduceOn* methods would be the ones with the Collectors. Basically, you'll just have to rename what you have already implemented. The reduceOn* methods will only return one value per vertex, like they do currently, but internally use the reduce operator (instead of reduceGroup).

For example, if you want to compute the sum of edge values, you can use the `reduceOnEdges` method, same as you currently do. And if you want to compute the Jaccard coefficient, you need to output multiple values per vertex and for that you can use `groupReduceOnEdges`.

Please let me know if you have any questions! Thanks again :-)

**Comment by andralungu** [comment]
Hello @vasia ,

I updated the PR to use reduce where possible(for the functions which return one value per vertex).
Let me know if there is anything I missed.

Thanks!

File: docs/gelly_guide.md [hunk 1]
```diff
@@ -269,7 +269,15 @@ Neighborhood Methods
...
+`groupReduceOnEdges()` can be used to compute an aggregation on the neighboring edges of a vertex,
+while `groupReduceOnNeighbors()` has access to both the neighboring edges and vertices. The neighborhood scope
+is defined by the `EdgeDirection` parameter, which takes the values `IN`, `OUT` or `ALL`. `IN` will gather all in-coming edges (neighbors) of a vertex, `OUT` will gather all out-going edges (neighbors), while `ALL` will gather all edges (neighbors).
+
+The `groupReduceOnEdges()` and `groupReduceOnNeighbors()` methods return zero, one or more values per vertex.
+When returning a single value per vertex, `reduceOnEdges()` or `reduceOnNeighbors()` should be called
+as they are more efficient. Nevertheless, when the reduce on edges modifies the value produced per vertex, for
+instance by multiplying it with a constant, `groupReduceOnEdges()` or `groupReduceOnNeighbors()` must be used
+as illustrated in the third code snippet.
```
Comment (vasia) [inline docs/gelly_guide.md]: I would rephrase this into something like the following: "when the user-defined function to be applied on the neighborhood is associative and commutative, it is highly advised to use the `reduceOnEdges()` and `reduceOnNeighbors()` methods. These methods can exploit combiners internally and significantly improve performance".

File: docs/gelly_guide.md [hunk 2]
```diff
@@ -282,25 +290,28 @@ The following code will collect the out-edges for each vertex and apply the `Sel
...
-}
+			for (Edge<Long, Long> edge: edges) {
+				if (edge.getValue() < weight) {
+					weight = edge.getValue();
+					minNeighborId = edge.getTarget();
+				}
+			}
+			out.collect(new Tuple2<Long, Long>(v.getId(), minNeighborId));
+		}
+	}
```
Comment (vasia) [inline docs/gelly_guide.md]: I think it's confusing what this example is doing now. The description says that each vertex selects the min weight among its edges, yet you modified it to return the ID of the target neighbor of that edge. Do you want to show something that I'm missing here?

Comment (vasia) [inline docs/gelly_guide.md]: Also, the description says `SelectMinWeight` but the method name is `SelectMinWeightNeighbor`.

File: docs/gelly_guide.md [hunk 3]
```diff
@@ -316,28 +327,49 @@ DataSet<Tuple2<Long, Long>> verticesWithSum = graph.reduceOnNeighbors(
...
+    	long sum = firstNeighbor.f2.getValue() + secondNeighbor.f2.getValue();
+    	return new Tuple3<Long, Edge<Long, Long>, Vertex<Long, Long>>(firstNeighbor.f0, firstNeighbor.f1,
+    			new Vertex<Long, Long>(firstNeighbor.f0, sum));
+    }
 }
 {% endhighlight %}
 
 <p class="text-center">
-    <img alt="reduseOnNeighbors Example" width="70%" src="img/gelly-reduceOnNeighbors.png"/>
+    <img alt="reduceOnNeighbors Example" width="70%" src="img/gelly-reduceOnNeighbors.png"/>
```
Comment (vasia) [inline docs/gelly_guide.md]: good catch :+1:

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/EdgeDirection.java [hunk 4]
```diff
@@ -20,10 +20,10 @@
 
 /**
  * The EdgeDirection is used to select a node's neighborhood
- * by the {@link Graph#reduceOnEdges(EdgesFunction, EdgeDirection)},
- * {@link Graph#reduceOnEdges(EdgesFunctionWithVertexValue, EdgeDirection)},
- * {@link Graph#reduceOnNeighbors(NeighborsFunction, EdgeDirection)} and
- * {@link Graph#reduceOnNeighbors(NeighborsFunctionWithVertexValue, EdgeDirection)}
+ * by the {@link Graph#groupReduceOnEdges(EdgesFunction, EdgeDirection)},
```
Comment (vasia) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/EdgeDirection.java]: isn't it used in both cases?

**Comment by vasia** [comment]
Hi @andralungu! Thanks for the quick update ^^
The result looks good, but I have a few more comments/suggestions (apart from the inline ones).

- +1 for the `groupReduce*` implementations. These look great :-)

- Regarding the combinable versions, I think that the current API is not very intuitive. I understand that you did it like this because you were restricted by the return type of the `ReduceFunction`. I think that it would be better if we simplify these to operate directly on the values, instead of tuples with nested Edge and Vertex data.
For example, say you want to compute the sum of all out-neighbors of a vertex. The value types to combine inside the reduce function should be the same as the vertex value type, instead of a Tuple3.
In the case of reducing on edges, the type should be the same as the edge value type.
Thus, I would make `reduceOnEdges` return `DataSet<Tuple2<K, EV>` and `reduceOnNeighbors` return `DataSet<Tuple2<K, VV>`, i.e. one value per vertex ID.

-  There is some inconsistency in the naming of the methods, for example the user-defined methods in the group case are called `iterateOn...` and in the reduce case they are called `reduce...`. We should stick to one of these :-) Also, maybe we could find a better name for `ReduceEdgesFunction` and `ReduceNeighborsFunction`. I don't have a suggestion right now, but I'll think about it.

-  Finally, there are several warnings in your code for unchecked and raw types. Could you turn on warning notifications in your IDE and try to add suppress annotations to get of them?

Let me know what you think and whether you have any questions!

Thanks again!

**Comment by andralungu** [comment]
Hi @vasia ,

The inline comments have been addressed, I still have the big comment to look at :)

One thing(or two ^^) that could be discussed: 
- the inconsistency between the iterateOn* and the reduce* method names comes from the fact that the two methods have different behaviour: reduce takes the elements two by two and aggregates them while iterateOn uses an iterator to go through the edges. Hence the different names. 
Imagine something like this:
Tuple2<K, Edge<K, EV>> iterateOnEdges(Tuple2<K, Edge<K, EV>> firstEdge, Tuple2<K, Edge<K, EV>> secondEdge);
when you're not iterating on anything(or vice-versa reduceOnEdges(Iterable edges)).

- I know ReduceEdgesFunction is not a pretty name, but I don't have another one, ReduceFunction is already taken :P. When you think of a better name, tell me and I can do the refactoring.

**Comment by andralungu** [comment]
Apart from the fact that I don't agree with those two methods having the same name, everything else was updated.

**Comment by vasia** [comment]
Hey @andralungu!
I see your point regarding `iterateOn...` and `reduce...`. Let's leave it like this for now.
Only small thing that maybe I didn't explain well, is "operate directly on the values" for the combinable versions. I meant there's no need to pass the vertex keys at all.
This is a small change, I can make it before merging.
Otherwise, everything looks good, I'll merge right after #408. Thanks!

**Pull request closed without merging**